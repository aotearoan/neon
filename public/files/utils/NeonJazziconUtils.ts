import { NeonRandomUtils } from './NeonRandomUtils';

export class NeonJazziconUtils {
  private static readonly shapeCount = 4;
  private static readonly svgns = 'http://www.w3.org/2000/svg';
  private static readonly wobble = 30;

  /**
   * Returns base-64 encoded data URI
   * @param palette the color palettes to use when generating the Jazzicon
   * @param name used for random seed
   * @param size size in pixels to generate
   */
  public static genSvg(palette: Array<string>, name: string, size: number): string {
    return NeonJazziconUtils.generateIdenticon(palette, size, NeonRandomUtils.rand(name)).outerHTML;
  }

  private static generateIdenticon(palette: Array<string>, diameter: number, rand: () => number): SVGElement {
    const remainingColors = NeonJazziconUtils.hueShift(palette.slice(), rand);

    const svg = document.createElementNS(NeonJazziconUtils.svgns, 'svg');
    svg.setAttribute('xmlns', NeonJazziconUtils.svgns);
    svg.setAttributeNS(null, 'x', '0');
    svg.setAttributeNS(null, 'y', '0');
    svg.setAttributeNS(null, 'width', `${diameter}`);
    svg.setAttributeNS(null, 'height', `${diameter}`);

    NeonJazziconUtils.genShape(
      [NeonJazziconUtils.genColor(remainingColors, rand)],
      diameter,
      0,
      NeonJazziconUtils.shapeCount - 1,
      svg,
      () => 0,
    );

    for (let i = 0; i < NeonJazziconUtils.shapeCount - 1; i++) {
      NeonJazziconUtils.genShape(remainingColors, diameter, i, NeonJazziconUtils.shapeCount - 1, svg, rand);
    }

    return svg;
  }

  private static genShape(
    remainingColors: string[],
    diameter: number,
    i: number,
    total: number,
    svg: SVGElement,
    rand: () => number,
  ) {
    const center = diameter / 2;

    const shape = document.createElementNS(NeonJazziconUtils.svgns, 'rect');
    shape.setAttributeNS(null, 'x', '0');
    shape.setAttributeNS(null, 'y', '0');
    shape.setAttributeNS(null, 'width', `${diameter}`);
    shape.setAttributeNS(null, 'height', `${diameter}`);

    const firstRot = rand();
    const angle = Math.PI * 2 * firstRot;
    const velocity = (diameter / total) * rand() + (i * diameter) / total;

    const tx = Math.cos(angle) * velocity;
    const ty = Math.sin(angle) * velocity;

    const translate = `translate(${tx} ${ty})`;

    // Third random is a shape rotation on top of all of that.
    const secondRot = rand();
    const rot = firstRot * 360 + secondRot * 180;
    const rotate = 'rotate(' + rot.toFixed(1) + ' ' + center + ' ' + center + ')';
    const transform = translate + ' ' + rotate;
    shape.setAttributeNS(null, 'transform', transform);
    const fill = NeonJazziconUtils.genColor(remainingColors, rand);
    shape.setAttributeNS(null, 'fill', fill);

    svg.appendChild(shape);
  }

  private static genColor(colors: string[], rand: () => number) {
    const idx = Math.floor(colors.length * rand());
    return colors.splice(idx, 1)[0];
  }

  private static hueShift(colors: Array<string>, rand: () => number) {
    const amount = rand() * 30 - NeonJazziconUtils.wobble / 2;
    const rotate = (hex: string) => NeonJazziconUtils.colorRotate(hex, amount);
    return colors.map(rotate);
  }

  private static colorRotate(hex: string, degrees: number) {
    const hsl = NeonJazziconUtils.hexToHSL(hex);
    let hue = hsl.h;
    hue = (hue + degrees) % 360;
    hue = hue < 0 ? 360 + hue : hue;
    hsl.h = hue;
    return NeonJazziconUtils.hslToHex(hsl);
  }

  private static hexToHSL(hex: string) {
    // Convert hex to RGB first
    let r = +('0x' + hex[1] + hex[2]);
    let g = +('0x' + hex[3] + hex[4]);
    let b = +('0x' + hex[5] + hex[6]);
    // Then to HSL
    r /= 255;
    g /= 255;
    b /= 255;
    const cmin = Math.min(r, g, b);
    const cmax = Math.max(r, g, b);
    const delta = cmax - cmin;
    let h = 0;
    let s = 0;
    let l = 0;

    if (delta === 0) {
      h = 0;
    } else if (cmax === r) {
      h = ((g - b) / delta) % 6;
    } else if (cmax === g) {
      h = (b - r) / delta + 2;
    } else {
      h = (r - g) / delta + 4;
    }

    h = Math.round(h * 60);

    if (h < 0) {
      h += 360;
    }

    l = (cmax + cmin) / 2;
    s = delta === 0 ? 0 : delta / (1 - Math.abs(2 * l - 1));
    s = +(s * 100).toFixed(1);
    l = +(l * 100).toFixed(1);

    return { h, s, l };
  }

  private static hslToHex(hsl: Record<string, number>) {
    const h = hsl.h;
    let { s, l } = hsl;
    s /= 100;
    l /= 100;

    const c = (1 - Math.abs(2 * l - 1)) * s;
    const x = c * (1 - Math.abs(((h / 60) % 2) - 1));
    const m = l - c / 2;
    let r = 0;
    let g = 0;
    let b = 0;

    if (h >= 0 && h < 60) {
      r = c;
      g = x;
      b = 0;
    } else if (h >= 60 && h < 120) {
      r = x;
      g = c;
      b = 0;
    } else if (h >= 120 && h < 180) {
      r = 0;
      g = c;
      b = x;
    } else if (h >= 180 && h < 240) {
      r = 0;
      g = x;
      b = c;
    } else if (h >= 240 && h < 300) {
      r = x;
      g = 0;
      b = c;
    } else if (h >= 300 && h < 360) {
      r = c;
      g = 0;
      b = x;
    }
    // Having obtained RGB, convert channels to hex
    let rStr = Math.round((r + m) * 255).toString(16);
    let gStr = Math.round((g + m) * 255).toString(16);
    let bStr = Math.round((b + m) * 255).toString(16);

    // Prepend 0s, if necessary
    if (rStr.length === 1) {
      rStr = '0' + r;
    }

    if (gStr.length === 1) {
      gStr = '0' + g;
    }

    if (bStr.length === 1) {
      bStr = '0' + b;
    }

    return '#' + rStr + gStr + bStr;
  }
}
