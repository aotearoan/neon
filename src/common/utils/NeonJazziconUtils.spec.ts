import { NeonJazziconUtils } from './NeonJazziconUtils';

describe('NeonJazziconUtils', () => {
  it('renders jazzicons', () => {
    let palette = ['#000000', '#666666', '#ffffff'];
    expect(NeonJazziconUtils.genSvg(palette, 'Test 1234', 32)).toMatchSnapshot();
    expect(NeonJazziconUtils.genSvg(palette, 'Test 123456789', 32)).toMatchSnapshot();
    palette = ['#ff0000', '#660066', '#ffff00'];
    expect(NeonJazziconUtils.genSvg(palette, 'Test 9876', 64)).toMatchSnapshot();
    expect(NeonJazziconUtils.genSvg(palette, 'Test 9876543', 64)).toMatchSnapshot();
    palette = ['#00ff00', '#000066', '#f0f0e0'];
    expect(NeonJazziconUtils.genSvg(palette, 'Another test 9876', 64)).toMatchSnapshot();
    expect(NeonJazziconUtils.genSvg(palette, 'Another test 9876543', 64)).toMatchSnapshot();
  });
});
