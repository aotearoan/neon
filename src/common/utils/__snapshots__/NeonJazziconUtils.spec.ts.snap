// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NeonJazziconUtils renders jazzicons 1`] = `
<svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="32" height="32">
  <rect x="0" y="0" width="32" height="32" transform="translate(0 0) rotate(0.0 16 16)" fill="#666666"></rect>
  <rect x="0" y="0" width="32" height="32" transform="translate(1.1409489889637223 6.439504089942269) rotate(224.4 16 16)" fill="#ffffff"></rect>
  <rect x="0" y="0" width="32" height="32" transform="translate(-1.2528046450621513 -21.241258963003112) rotate(313.2 16 16)" fill="#000000"></rect>
  <rect x="0" y="0" width="32" height="32" transform="translate(13.637224070387488 24.889953777971208) rotate(144.4 16 16)" fill="undefined"></rect>
</svg>
`;

exports[`NeonJazziconUtils renders jazzicons 2`] = `
<svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="32" height="32">
  <rect x="0" y="0" width="32" height="32" transform="translate(0 0) rotate(0.0 16 16)" fill="#000000"></rect>
  <rect x="0" y="0" width="32" height="32" transform="translate(-6.718492075605844 2.3090033120392337) rotate(180.9 16 16)" fill="#666666"></rect>
  <rect x="0" y="0" width="32" height="32" transform="translate(-0.2225313310323077 -16.740563608535048) rotate(274.7 16 16)" fill="#ffffff"></rect>
  <rect x="0" y="0" width="32" height="32" transform="translate(9.981358448505167 -26.404615582276236) rotate(440.8 16 16)" fill="undefined"></rect>
</svg>
`;

exports[`NeonJazziconUtils renders jazzicons 3`] = `
<svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="64" height="64">
  <rect x="0" y="0" width="64" height="64" transform="translate(0 0) rotate(0.0 32 32)" fill="#ff1800"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(0.5394866348947869 -2.4291338427200637) rotate(371.7 32 32)" fill="#e7ff00"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(36.11125979293032 -22.149669364154377) rotate(487.3 32 32)" fill="#66005c"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(32.015699249238175 -52.37192969784017) rotate(371.7 32 32)" fill="undefined"></rect>
</svg>
`;

exports[`NeonJazziconUtils renders jazzicons 4`] = `
<svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="64" height="64">
  <rect x="0" y="0" width="64" height="64" transform="translate(0 0) rotate(0.0 32 32)" fill="#ff00.0361822783015668400"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(0.46849290721454095 6.442424109267294) rotate(148.0 32 32)" fill="#f6ff00"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(10.270409110358345 31.663375166094795) rotate(147.9 32 32)" fill="#660062"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(42.60843338187566 15.987876672655743) rotate(172.9 32 32)" fill="undefined"></rect>
</svg>
`;

exports[`NeonJazziconUtils renders jazzicons 5`] = `
<svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="64" height="64">
  <rect x="0" y="0" width="64" height="64" transform="translate(0 0) rotate(0.0 32 32)" fill="#f0ede0"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(-2.447986482966687 11.373404163400103) rotate(127.0 32 32)" fill="#35ff00"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(7.249870688857548 -33.459996643863185) rotate(458.2 32 32)" fill="#001566"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(12.438852936747866 -43.873659459047886) rotate(455.4 32 32)" fill="undefined"></rect>
</svg>
`;

exports[`NeonJazziconUtils renders jazzicons 6`] = `
<svg xmlns="http://www.w3.org/2000/svg" x="0" y="0" width="64" height="64">
  <rect x="0" y="0" width="64" height="64" transform="translate(0 0) rotate(0.0 32 32)" fill="#eef0e0"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(2.920421620956669 4.172687764328594) rotate(166.6 32 32)" fill="#00.0600846708286553660066"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(-9.435688998454925 -28.95825177650139) rotate(355.0 32 32)" fill="#00ff26"></rect>
  <rect x="0" y="0" width="64" height="64" transform="translate(36.0595495198541 38.37936461054924) rotate(139.9 32 32)" fill="undefined"></rect>
</svg>
`;
